symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+                if resp.status == 200 and (
+                    (
+                        "microsoft" in server
+                        and resp.headers["Content-Type"] == "application/octet-stream"
+                    )
+                    or "mozilla" in server
+                ):
+                    log.debug(f"File exists: {url}")
+                    return True
+                else:
+                    return False
+        except Exception as e:
+            log.debug(f"Error with {url}: retry")
+            log.exception(e)
+            await asyncio.sleep(0.5)
+
+    log.debug(f"Too much retries (HEAD) for {url}: give up.")
+    return False
+
+
+async def fetch_file(client, server, filename):
+    """
+    Fetch the file from the server
+    """
+    url = urljoin(server, quote(filename))
+    log.debug(f"Fetch url: {url}")
+    for _ in range(10):
+        try:
+            async with client.get(url, headers=HEADERS, allow_redirects=True) as resp:
+                if resp.status == 200:
+                    data = await resp.read()
+                    typ = get_type(data)
+                    if typ == "unknown":
+                        # try again
+                        await asyncio.sleep(0.5)
+                    elif typ == "pdb-v2":
+                        # too old: skip it
+                        log.debug(f"PDB v2 (skipped because too old): {url}")
+                        return None
                     else:
-                        log.warn("Couldn't fetch binary for %s, %s",
-                                 code_id, code_file)
-                        raise Win64ProcessError()
-                    continue
+                        return data
                 else:
-                    raise Win64ProcessError()
-            return None
+                    log.error(f"Cannot get data (status {resp.status}) for {url}: ")
+        except Exception as e:
+            log.debug(f"Error with {url}")
+            log.exception(e)
+            await asyncio.sleep(0.5)
 
-
-def server_has_file(filename):
-    '''
-    Send the symbol server a HEAD request to see if it has this symbol file.
-    '''
-    try:
-        r = requests.head(
-            urlparse.urljoin(
-                MOZILLA_SYMBOL_SERVER,
-                urllib.quote(filename)))
-        return r.status_code == 200
-    except requests.exceptions.RequestException:
-        return False
+    log.debug(f"Too much retries (GET) for {url}: give up.")
+    return None
 
 
 def write_skiplist(skiplist):
-    try:
-        with open(os.path.join(thisdir, 'skiplist.txt'), 'w') as sf:
-            for (debug_id, debug_file) in skiplist.iteritems():
-                sf.write('%s %s\n' % (debug_id, debug_file))
-    except IOError:
-        log.exception('Error writing skiplist.txt')
+    with open(os.path.join(thisdir, "skiplist.txt"), "w") as sf:
+        for (debug_id, debug_file) in skiplist.items():
```

Comment:
Nit: `sf.writelines("%s %s\n" % item for item in skiplist.items())`

Reply:
Actually since this is now Python 3, we can use f-strings:
```Python
sf.writelines(f"{debug_id} {debug_file}\n" for debug_id, debug_file in skiplist.items())
```